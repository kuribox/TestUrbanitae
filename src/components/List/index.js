import React from 'react'
import Loader from "react-loader-spinner";
import styles from './list.module.scss';
import ListElement from "../../components/ListElement";

const ListContainer = ({ loading, searchKey, elements, errorMsg, isDirty, onPress}) => {
  return (
    <div className={styles.listContent}>
      {loading && (
        <div data-testid="loading">
          <Loader
            type="Puff"
            color="#88b04b"
            height={30}
            width={30}
            />
        </div>
      )}

      {
        !loading && errorMsg && isDirty && (
        <div className={styles.errorMsg} data-testid="error"> 
            <span>{errorMsg}</span>
        </div>
        )
      }

      {
        !loading && !errorMsg && elements && elements.length === 0 && isDirty && (
        <div className={styles.errorMsg} data-testid="noResults">
            <span>No se encontraron resultados para su busqueda</span>
        </div>
        )
      }

      {
        !loading && elements && !errorMsg && elements.length > 0 && isDirty && (
          <>
            <div className={styles.searchTitle}>
                Resultados de la busqueda de {searchKey}
            </div>
            <div className={styles.results} data-testid="results">
              {
                elements.map(e => <ListElement key={e.id} element={e} onPress={onPress} />)
              }
            </div>
          </>
        )
      }
    </div>
  );
}

export default ListContainer;
