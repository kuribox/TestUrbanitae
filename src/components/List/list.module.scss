.errorMsg {
    margin-top: 15px;
    margin-bottom: 15px;
    background-color: lighten(red, 40%);
    border: 3px solid lighten(red, 20%);;
    border-radius: 5px;
    padding: 10px;
    justify-content: center;
    align-items: center;
    display: flex;
    color: lighten(red, 20%);
}

.listContent {
    margin-top: 15px;
    flex: 1;
    display: flex;
    flex-direction: column;
    width: 100%;
    align-items: center;
}

.searchTitle {
    font-weight: bold;
    font-size: 14px;
    margin-bottom: 10px;
}

.results {
    width: 100%;
    display: flex;
    flex-direction: column;
}