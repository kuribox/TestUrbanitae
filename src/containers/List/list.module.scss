.container {
    display: flex;
    flex: 1;
    flex-direction: column;
    justify-content: flex-start;
    align-items: center;
}

.errorMsg {
    margin-top: 15px;
    margin-bottom: 15px;
    background-color: lighten(red, 40%);
    border: 3px solid lighten(red, 20%);;
    border-radius: 5px;
    padding: 10px;
    justify-content: center;
    align-items: center;
    display: flex;
    color: lighten(red, 20%);
}

.listContent {
    margin-top: 15px;
    flex: 1;
    display: flex;
    flex-direction: column;
    width: 100%;
    align-items: center;
}

.searchTitle {
    font-weight: bold;
    font-size: 14px;
    margin-bottom: 10px;
}