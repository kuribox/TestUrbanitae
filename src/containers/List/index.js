import React, { useState } from 'react'
import { useSelector, useDispatch } from 'react-redux';
import styles from './list.module.scss';
import { searchTitleAction } from '../../actions';
import SearchBox from '../../components/SearchBox';
import List from '../../components/List';
import { useHistory } from 'react-router-dom';

const ListContainer = () => {
  const history = useHistory();

  const [timeoutToSearch, setTimeoutToSearch] = useState(null);
  const [isDirty, setIsDirty] = useState(false);

  const loading = useSelector((state) => state.searchReducer.loading);
  const searchKey = useSelector((state) => state.searchReducer.searchKey);
  const results = useSelector((state) => state.searchReducer.results);
  const errorMsg = useSelector((state) => state.searchReducer.error);

  const dispatch = useDispatch();

  return (
    <div className={styles.container}>
      <SearchBox 
        onChange={(event) => {
          // Calcula el timeout antes de iniciar con la busqueda
          if (timeoutToSearch) clearTimeout(timeoutToSearch);
          setTimeoutToSearch(setTimeout(() => {
            dispatch(searchTitleAction({ value: event.target.value}));
            setIsDirty(true);
          }, 500));
        }}
      />

      <List 
        loading={loading}
        searchKey={searchKey}
        elements={results}
        errorMsg={errorMsg}
        isDirty={isDirty}
        onPress={(id) => {
          history.push(`detail/${id}`)
        }}
      />
    </div>
  );
}

export default ListContainer;
