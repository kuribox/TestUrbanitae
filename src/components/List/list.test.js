import React from 'react';
import renderer from 'react-test-renderer';
import {cleanup, fireEvent, render, screen} from '@testing-library/react';
import List from './index.js';

test('Snapshot check', () => {
  const component = renderer.create(
    <List />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

afterEach(cleanup);

test('Check init State', () => {
  const { queryByTestId, queryByText } =  render(
    <List 
      loading={false}
    />,
  );

  expect(queryByTestId('error')).toBeNull();
  expect(queryByText("No se encontraron resultados para su busqueda")).toBeNull();
  expect(queryByTestId('noResults')).toBeNull();
  expect(queryByTestId('results')).toBeNull();
});


test('Check no results', () => {
  const { queryByTestId, queryByText } = render(
    <List 
      loading={false}
      searchKey={"test"}
      elements={[]}
      errorMsg={null}
      isDirty={true}
    />,
  );

  expect(queryByTestId('error')).toBeNull();

  expect(queryByText("No se encontraron resultados para su busqueda")).toBeInTheDocument();
  expect(queryByTestId('noResults')).toBeInTheDocument();
  expect(queryByTestId('results')).toBeNull();
});

test('Check error message', () => {
  const { queryByTestId, queryByText } = render(
    <List 
      loading={false}
      searchKey={"test"}
      elements={[]}
      errorMsg="ERROR MESSAGE"
      isDirty={true}
    />,
  );

  expect(queryByTestId('error')).toBeInTheDocument();
  expect(queryByText("ERROR MESSAGE")).toBeInTheDocument();

  expect(queryByText("No se encontraron resultados para su busqueda")).toBeNull();
  expect(queryByTestId('noResults')).toBeNull();

  expect(queryByTestId('results')).toBeNull();
});


test('Check no error with no dirty list', () => {
  const { queryByTestId, queryByText } = render(
    <List 
      loading={false}
      searchKey={"test"}
      elements={[]}
      errorMsg="ERROR MESSAGE"
      isDirty={false}
    />,
  );

  expect(queryByTestId('error')).toBeNull();
  expect(queryByText("ERROR MESSAGE")).toBeNull();

  expect(queryByText("No se encontraron resultados para su busqueda")).toBeNull();
  expect(queryByTestId('noResults')).toBeNull();

  expect(queryByTestId('results')).toBeNull();
});


test('Check results list', () => {
  const { queryByTestId, queryByText } = render(
    <List 
      loading={false}
      searchKey={"test"}
      elements={[ { id: 1 }, { id: 2 } ]}
      errorMsg={null}
      isDirty={true}
    />,
  );

  expect(queryByTestId('error')).toBeNull();
  expect(queryByText("ERROR MESSAGE")).toBeNull();

  expect(queryByText("No se encontraron resultados para su busqueda")).toBeNull();
  expect(queryByTestId('noResults')).toBeNull();

  expect(queryByTestId('results')).toBeInTheDocument();
});

test('Check loading', () => {
  const { queryByTestId, queryByText } = render(
    <List 
      loading={true}
      searchKey={"test"}
      elements={[ { id: 1 }, { id: 2 } ]}
      errorMsg={null}
      isDirty={true}
    />,
  );

  expect(queryByTestId('error')).toBeNull();
  expect(queryByText("ERROR MESSAGE")).toBeNull();

  expect(queryByText("No se encontraron resultados para su busqueda")).toBeNull();
  expect(queryByTestId('noResults')).toBeNull();

  expect(queryByTestId('results')).toBeNull();
  expect(queryByTestId('loading')).toBeInTheDocument();
});

test('Check loading with Error', () => {
  const { queryByTestId, queryByText } = render(
    <List 
      loading={true}
      searchKey={"test"}
      elements={[ { id: 1 }, { id: 2 } ]}
      errorMsg="ERROR MESSAGE"
      isDirty={true}
    />,
  );

  expect(queryByTestId('error')).toBeNull();
  expect(queryByText("ERROR MESSAGE")).toBeNull();

  expect(queryByText("No se encontraron resultados para su busqueda")).toBeNull();
  expect(queryByTestId('noResults')).toBeNull();

  expect(queryByTestId('results')).toBeNull();

  expect(queryByTestId('loading')).toBeInTheDocument();
});

test('Check click element', () => {
  const handleClick = jest.fn();

  const { queryByTestId } = render(
    <List 
      loading={false}
      searchKey={"test"}
      elements={[ { id: 1 }, { id: 2 } ]}
      isDirty={true}
      onPress={handleClick}
    />,
  );

  expect(queryByTestId('results')).toBeInTheDocument();

  fireEvent.click(queryByTestId('results').firstChild)
});