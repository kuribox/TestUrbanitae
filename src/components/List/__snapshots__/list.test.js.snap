// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot check 1`] = `
<div
  className="listContent"
/>
`;
